File: backend/app/schemas/student.py
from datetime import date, datetime
from typing import List, Optional
from pydantic import BaseModel, EmailStr


# ── Student Profile ───────────────────────────────────────────────────────────

class StudentProfileUpdate(BaseModel):
    full_name: Optional[str] = None
    phone_number: Optional[str] = None
    alternate_email: Optional[EmailStr] = None
    date_of_birth: Optional[date] = None
    gender: Optional[str] = None
    address: Optional[str] = None
    current_year: Optional[int] = None
    current_semester: Optional[int] = None


class StudentProfileOut(BaseModel):
    id: str
    user_id: str
    full_name: str
    roll_number: str
    department_id: str
    current_year: int
    current_semester: int
    date_of_birth: Optional[date] = None
    gender: Optional[str] = None
    phone_number: Optional[str] = None
    alternate_email: Optional[str] = None
    address: Optional[str] = None
    profile_photo_url: Optional[str] = None
    profile_completion_percentage: float
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# ── Education ─────────────────────────────────────────────────────────────────

class EducationCreate(BaseModel):
    education_level: str  # 10TH | 12TH | DIPLOMA
    percentage: Optional[float] = None
    cgpa: Optional[float] = None
    board: Optional[str] = None
    passing_year: Optional[int] = None


class EducationOut(EducationCreate):
    id: str
    student_id: str
    model_config = {"from_attributes": True}


# ── Academic Records ──────────────────────────────────────────────────────────

class AcademicRecordCreate(BaseModel):
    semester: int
    cgpa: Optional[float] = None
    active_backlogs: int = 0
    total_backlogs_cleared: int = 0


class AcademicRecordOut(AcademicRecordCreate):
    id: str
    student_id: str
    model_config = {"from_attributes": True}


# ── Skills ────────────────────────────────────────────────────────────────────

class SkillCreate(BaseModel):
    name: str
    category: str = "TECHNICAL"


class SkillOut(BaseModel):
    id: str
    name: str
    category: str
    model_config = {"from_attributes": True}


class StudentSkillCreate(BaseModel):
    skill_name: str          # auto-creates skill if not exists
    category: str = "TECHNICAL"
    proficiency_level: Optional[str] = None


class StudentSkillOut(BaseModel):
    id: str
    skill: SkillOut
    proficiency_level: Optional[str] = None
    model_config = {"from_attributes": True}


# ── Projects ──────────────────────────────────────────────────────────────────

class ProjectCreate(BaseModel):
    title: str
    description: Optional[str] = None
    project_link: Optional[str] = None
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    technologies: List[str] = []   # list of skill names


class ProjectOut(BaseModel):
    id: str
    title: str
    description: Optional[str] = None
    project_link: Optional[str] = None
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    technologies: List[SkillOut] = []
    model_config = {"from_attributes": True}


# ── Experience ────────────────────────────────────────────────────────────────

class ExperienceCreate(BaseModel):
    company_name: str
    role_title: str
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    description: Optional[str] = None
    currently_working: bool = False


class ExperienceOut(ExperienceCreate):
    id: str
    student_id: str
    model_config = {"from_attributes": True}


# ── Certifications ────────────────────────────────────────────────────────────

class CertificationCreate(BaseModel):
    title: str
    issuing_organization: Optional[str] = None
    date_issued: Optional[date] = None
    credential_url: Optional[str] = None


class CertificationOut(CertificationCreate):
    id: str
    student_id: str
    model_config = {"from_attributes": True}


# ── Preferences ───────────────────────────────────────────────────────────────

class PreferencesUpdate(BaseModel):
    preferred_company_type: Optional[str] = None
    expected_package: Optional[float] = None
    willing_to_relocate: Optional[bool] = None
    preferred_roles: Optional[List[str]] = None      # job role names
    preferred_locations: Optional[List[str]] = None  # city names


class PreferencesOut(BaseModel):
    preferred_company_type: str
    expected_package: Optional[float] = None
    willing_to_relocate: bool
    preferred_roles: List[str] = []
    preferred_locations: List[str] = []
    model_config = {"from_attributes": True}


# ── Full profile (combined) ───────────────────────────────────────────────────

class FullProfileOut(BaseModel):
    profile: StudentProfileOut
    education: List[EducationOut] = []
    academic_records: List[AcademicRecordOut] = []
    skills: List[StudentSkillOut] = []
    projects: List[ProjectOut] = []
    experiences: List[ExperienceOut] = []
    certifications: List[CertificationOut] = []
    preferences: Optional[PreferencesOut] = None

File: frontend/src/api/student.ts
import client from './client'

// ── Types ─────────────────────────────────────────────────────────────────────

export interface StudentProfileOut {
  id: string; user_id: string; full_name: string; roll_number: string
  department_id: string; current_year: number; current_semester: number
  date_of_birth?: string; gender?: string; phone_number?: string
  alternate_email?: string; address?: string; profile_photo_url?: string
  profile_completion_percentage: number; created_at: string; updated_at: string
}

export interface SkillOut { id: string; name: string; category: string }

export interface StudentSkillOut { id: string; skill: SkillOut; proficiency_level?: string }

export interface ProjectOut {
  id: string; title: string; description?: string; project_link?: string
  start_date?: string; end_date?: string; technologies: SkillOut[]
}

export interface ExperienceOut {
  id: string; student_id: string; company_name: string; role_title: string
  start_date?: string; end_date?: string; description?: string; currently_working: boolean
}

export interface CertificationOut {
  id: string; student_id: string; title: string; issuing_organization?: string
  date_issued?: string; credential_url?: string
}

export interface EducationOut {
  id: string; student_id: string; education_level: string
  percentage?: number; cgpa?: number; board?: string; passing_year?: number
}

export interface AcademicRecordOut {
  id: string; student_id: string; semester: number; cgpa?: number
  active_backlogs: number; total_backlogs_cleared: number
}

export interface PreferencesOut {
  preferred_company_type: string; expected_package?: number
  willing_to_relocate: boolean; preferred_roles: string[]; preferred_locations: string[]
}

export interface FullProfileOut {
  profile: StudentProfileOut
  education: EducationOut[]
  academic_records: AcademicRecordOut[]
  skills: StudentSkillOut[]
  projects: ProjectOut[]
  experiences: ExperienceOut[]
  certifications: CertificationOut[]
  preferences?: PreferencesOut
}

// ── API calls ─────────────────────────────────────────────────────────────────

export const studentApi = {
  getFullProfile: () =>
    client.get<FullProfileOut>('/student/profile'),

  updateProfile: (data: Partial<StudentProfileOut>) =>
    client.patch<StudentProfileOut>('/student/profile', data),

  upsertEducation: (items: Omit<EducationOut, 'id' | 'student_id'>[]) =>
    client.put<EducationOut[]>('/student/profile/education', items),

  upsertAcademicRecords: (items: Omit<AcademicRecordOut, 'id' | 'student_id'>[]) =>
    client.put<AcademicRecordOut[]>('/student/profile/academic-records', items),

  syncSkills: (items: { skill_name: string; category?: string; proficiency_level?: string }[]) =>
    client.put<StudentSkillOut[]>('/student/profile/skills', items),

  syncProjects: (items: { title: string; description?: string; project_link?: string; start_date?: string; end_date?: string; technologies: string[] }[]) =>
    client.put<ProjectOut[]>('/student/profile/projects', items),

  syncExperiences: (items: Omit<ExperienceOut, 'id' | 'student_id'>[]) =>
    client.put<ExperienceOut[]>('/student/profile/experience', items),

  syncCertifications: (items: Omit<CertificationOut, 'id' | 'student_id'>[]) =>
    client.put<CertificationOut[]>('/student/profile/certifications', items),

  updatePreferences: (data: Partial<PreferencesOut> & { preferred_roles?: string[]; preferred_locations?: string[] }) =>
    client.put<PreferencesOut>('/student/profile/preferences', data),
}
